### PR TITLE
Kullback-Leibler divergence

### PR DESCRIPTION
**Pull request by Atticus1806**
This is a draft for the Kullback-Leibler divergence (loss).

The idea is to use this with `.mark_as_loss()`. This is only a draft since I am not 100% sure about all the modalities, which might need to be taken care of (or are already dealt with with the marking)

In my mind this is: 
- conversion to actual (batch-wise) loss
- potential inconsistencies / stability concerns with the input e.g. PyTorch (https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html) assumes `estimated` to be in the log-space already and makes it optional for `target` to be too. Do we want to enforce/handle this in the same way?

### CODE REVIEW

**Comment by albertz** [comment]
It should be consistent to `cross_entropy` regarding the logic of log-space etc.

Also see the discussion in #38.

I don't think it should also perform the reduction over time or other dims. It should just calculate the KL. The final reduction and accumulation over time and batch would be handled by RETURNN.

**Comment by Atticus1806** [comment]
Okay I added the checks for the type of `estimated` the same way it is done for `ce`. So I think this should be ready then.

File: nn/loss.py [hunk 1]
```diff
@@ -57,18 +57,32 @@ def cross_entropy(*, target: nn.Tensor, estimated: nn.Tensor, estimated_type: st
   return -nn.dot(target, log_prob, reduce=axis)
 
 
-def kl_div(*, target: nn.Tensor, estimated: nn.Tensor):
+def kl_div(*, target: nn.Tensor, estimated: nn.Tensor, estimated_type: str, axis: Optional[nn.Dim] = None) -> nn.Tensor:
```
[the remark this answers is not in the document]
Comment (Atticus1806) [reply nn/loss.py:60]: I thought about this too. Only log space would not work here, since we need both log and probs. Only check would maybe be whether logits or probs. 
This isn't done for `ce` either, so maybe we should leave it like this here and maybe have one common PR change it?

[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:60]: Why does log space not work? You can simply do `exp(log_prob)` to get the probs?

But you are right, for CE we also don't have it. So maybe it's fine for now and this can be an extension later whenever we need it.

Note however that it can be inefficient this way because it might compute softmax twice. The `safe_log` on a previous softmax will convert it into a `log_softmax`, so then it calculates both the softmax and the log_softmax. It would be more efficient to only calculate log_softmax and then do just `exp` on the log probs.

[the remark this answers is not in the document]
Comment (Atticus1806) [reply nn/loss.py:60]: >You can simply do exp(log_prob) to get the probs?

If we assume that its always log to base e yes. I guess other bases are too uncommon to be used so its fine?

[the remark this answers is not in the document]
Comment (Atticus1806) [reply nn/loss.py:60]: >It would be more efficient to only calculate log_softmax and then do just exp on the log probs.

So you mean like this: `kl = nn.exp(log_target) * (log_target - log_est)` ? Or where would be calc. softmax twice?

[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:60]: > > It would be more efficient to only calculate log_softmax and then do just exp on the log probs.
> 
> So you mean like this: `kl = nn.exp(log_target) * (log_target - log_est)` ? Or where would be calc. softmax twice?

In `kl = target * (nn.safe_log(target) - log_est)`, we would calc softmax twice for target: `target = softmax(...)` and `nn.safe_log(target) = log_softmax(...)`.

[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:60]: Well, I'm not sure what's better (from a stability point of view). Doing exp(log_softmax(...)) vs softmax(...).

[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:60]: Also, despite "better" (both under efficiency consideration, and stability), what's more "straightforward", i.e. what would the user expect what would happen here?

[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:60]: So, maybe leave it like it is for now?

[the remark this answers is not in the document]
Comment (Atticus1806) [reply nn/loss.py:60]: As long as its fine in Efficiency, I think as it is its more readable and closer to the formula. So if you say its fine, I would leave it like this.

[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:60]: Well, as said, it's a bit suboptimal wrt efficiency. The current code results in softmax twice. The question is, what is more important, efficiency or maybe straightforwardness?

Maybe efficiency is actually more important because that is in general for RETURNN the policy.

[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:60]: See my suggested code change.

File: nn/loss.py [hunk 2]
```diff
@@ -55,3 +55,34 @@ def cross_entropy(*, target: nn.Tensor, estimated: nn.Tensor, estimated_type: st
...
+    log_est = nn.safe_log(estimated)
+  elif estimated_type == "log-probs":
+    log_est = estimated
+  elif estimated_type == "logits":
+    log_est = nn.log_softmax(estimated, axis=axis)
+  else:
+    raise ValueError("estimated_kind must be 'probs', 'log-probs' or 'logits'")
+
+  log_target = nn.safe_log(target)
+  kl = target * (log_target - log_est)
```
[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:86]: Ah, another thing: Shouldn't this be `nn.dot`, like in `cross_entropy`?

File: nn/loss.py [hunk 3]
```diff
@@ -55,3 +55,34 @@ def cross_entropy(*, target: nn.Tensor, estimated: nn.Tensor, estimated_type: st
   else:
     raise ValueError("estimated_kind must be 'probs', 'log-probs' or 'logits'")
   return -nn.dot(target, log_prob, reduce=axis)
+
+
+def kl_div(*, target: nn.Tensor, estimated: nn.Tensor, estimated_type: str, axis: Optional[nn.Dim] = None) -> nn.Tensor:
```
[the remark this answers is not in the document]
Comment (albertz) [reply nn/loss.py:60]: See my new code suggestion.

File: nn/loss.py [hunk 4]
```diff
@@ -55,3 +55,53 @@ def cross_entropy(*, target: nn.Tensor, estimated: nn.Tensor, estimated_type: st
...
+  elif estimated_type == "logits":
+    log_est = nn.log_softmax(estimated, axis=axis)
+  else:
+    raise ValueError("estimated_kind must be 'probs', 'log-probs' or 'logits'")
+
+  # Assuming target = softmax(...):
+  # Using nn.exp(log_target) instead of target (but not nn.safe_exp!)
+  # to avoid calculating softmax twice (efficiency)
+  # (because nn.safe_log(target) = log_softmax(...), so a separate softmax calculation).
+  log_target = nn.safe_log(target)
```
Comment (albertz) [inline nn/loss.py:104]:
```suggestion
```
